theano/gof/op.py

```diff
@@ -741,13 +741,24 @@ def rval():
         # condition: either there was no c_code, or it failed
 
         p = node.op.perform
-        # default arguments are stored in the closure of `rval`
 
-        def rval(p=p, i=node_input_storage, o=node_output_storage, n=node):
-            r = p(n, [x[0] for x in i], o)
-            for o in node.outputs:
-                compute_map[o][0] = True
-            return r
+        ctx = node.run_context()
+
+        if ctx is graph.NoContext:
+            # default arguments are stored in the closure of `rval`
+            def rval(p=p, i=node_input_storage, o=node_output_storage, n=node):
+                r = p(n, [x[0] for x in i], o)
+                for o in node.outputs:
+                    compute_map[o][0] = True
+                return r
+        else:
+            ctx_val = node.context_type.filter(ctx)
+            def rval(p=p, i=node_input_storage, o=node_output_storage, n=node,
+                     ctx=ctx_val):
+                r = p(n, [x[0] for x in i], o, ctx)
+                for o in node.outputs:
+                    compute_map[o][0] = True
+                return r
```

Comment:
The one in scan is fine since it will create thunks for scan itself which doesn't have a context.

This one had to change because it replace Op.make_thunk() which fiddles with arbitrary nodes.